pkg/bucketaccessrequest/bucketaccessrequest.go

```diff
@@ -0,0 +1,141 @@
+	b.kubeClient = k
+}
+
+func (b *bucketAccessRequestListener) InitializeBucketClient(bc bucketclientset.Interface) {
+	b.bucketClient = bc
+}
+
+func (b *bucketAccessRequestListener) Add(ctx context.Context, obj *v1alpha1.BucketAccessRequest) error {
+	glog.V(1).Infof("add called for bucketaccessrequest %s", obj.Name)
+	bucketAccessRequest := obj
+
+	err := b.provisionBucketAccess(ctx, bucketAccessRequest)
+	if err != nil {
+		// Provisioning is 100% finished / not in progress.
+		switch err {
+		case util.ErrInvalidBucketAccessClass:
+			glog.V(5).Infof("Bucket access class specified does not exist. Stop provisioning, removing bucket access request %s from bucket access requests in progress", bucketAccessRequest.UID)
+			err = nil
+		case util.ErrBucketAccessAlreadyExists:
+			glog.V(5).Infof("Bucket access already exist for this bucket request. Stop provisioning, removing bucket access request %s from bucket access requests in progress", bucketAccessRequest.UID)
+			err = nil
+		default:
+			glog.V(2).Infof("Final error received, removing bucket access request %s from bucket access requests in progress", bucketAccessRequest.Name)
+		}
+		return err
+	}
+
+	glog.V(5).Infof("BucketAccessRequest processing succeeded, removing bucket access request %s from bucket access requests in progress", bucketAccessRequest.Name)
+	return nil
+}
+
+func (b *bucketAccessRequestListener) Update(ctx context.Context, old, new *v1alpha1.BucketAccessRequest) error {
+	glog.V(1).Infof("update called for bucket %v", old)
+	return nil
+}
+
+func (b *bucketAccessRequestListener) Delete(ctx context.Context, obj *v1alpha1.BucketAccessRequest) error {
+	glog.V(1).Infof("delete called for bucket %v", obj)
+	return nil
+}
+
+// provisionBucketAccess  attempts to provision a BucketAccess for the given bucketAccessRequest.
+// Returns nil error only when the bucketaccess was provisioned. An error is return if we cannot create bucket access.
+// A normal error is returned when  bucket acess  was not provisioned and provisioning should be retried (requeue the bucketAccessRequest),
+// or a special error  errBucketAccessAlreadyExists, errInvalidBucketAccessClass is returned when provisioning was impossible and
+// no further attempts to provision should be tried.
+func (b *bucketAccessRequestListener) provisionBucketAccess(ctx context.Context, bucketAccessRequest *v1alpha1.BucketAccessRequest) error {
+	bucketAccessClassName := bucketAccessRequest.Spec.BucketAccessClassName
+
+	bucketaccess := b.FindBucketAccess(ctx, bucketAccessRequest)
+	if bucketaccess != nil {
+		// bucketaccess has been already provisioned, nothing to do.
+		return util.ErrBucketAccessAlreadyExists
+	}
+
+	bucketAccessClass, err := b.bucketClient.ObjectstorageV1alpha1().BucketAccessClasses().Get(ctx, bucketAccessClassName, metav1.GetOptions{})
+	if bucketAccessClass == nil {
+		// bucket has been already provisioned, nothing to do.
```

Comment:
If BAC is nil then the BAR did not specify the BAC or the BAC has been deleted. Why is the conclusion that the bucket has already been provisioned?

Reply:
Fixed the comment.